File: node-root/snake-ai/scripts/performanceTest/singleThread.ts
import * as fs from "fs";
import * as path from "path";
import SnakeGame from "snake-game/SnakeGame";
import InputLayer from "../../InputLayer";
import SnakeBrain from "../../SnakeBrain";
import { TimingUtils } from "./Timer";
import type { ISnakeBrain } from "../../SnakeBrain";

interface Result {
  score: number;
  execTime: number;
}

async function main() {
  const brainDataPath = path.join(__dirname, "trained-brain.json");
  const brainData: ISnakeBrain = JSON.parse(fs.readFileSync(brainDataPath, "utf8"));

  const snakeGame = new SnakeGame({ worldWidth: 20, worldHeight: 20 });
  const inputLayer = new InputLayer(snakeGame);
  const snakeBrain = new SnakeBrain({
    inputLength: brainData.inputLength,
    layerShapes: brainData.layerShapes,
    hiddenLayerActivationFunction: brainData.hiddenLayerActivationFunction,
    providedWeightsAndBiases: {
      weights: brainData.weights,
      biases: brainData.biases,
    },
  });

  const games = 1000;
  const result: Result[] = [];

  for (let i = 0; i < games; i++) {
    snakeGame.reset();

    const execTime = await TimingUtils.execTime(async () => {
      while (!snakeGame.gameOver) {
        const input = inputLayer.compute();
        const direction = snakeBrain.compute(input);
        snakeGame.snakeMoveByDirection(direction);
      }
    });

    result.push({
      score: snakeGame.snake.length,
      execTime,
    });
  }

  const totalTime = result.reduce((acc, curr) => acc + curr.execTime, 0);
  console.log(`Total time: ${totalTime.toFixed(3)}s`);

  const bestScore = result.reduce((acc, curr) => Math.max(acc, curr.score), 0);
  console.log(`Best score: ${bestScore}`);

  const avgScore = result.reduce((acc, curr) => acc + curr.score, 0) / games;
  console.log(`Average score: ${avgScore.toFixed(2)}`);
}

main();

File: kotlin-root/gasnake/src/main/kotlin/com/example/snake/ai/performance/SingleThread.kt
package com.example.snake.ai.performance


import com.example.snake.ai.InputLayer
import com.example.snake.ai.ProvidedWeightsAndBiases
import com.example.snake.ai.SnakeBrain
import com.example.snake.ai.SnakeBrainData
import com.example.snake.game.Options
import com.example.snake.game.SnakeGame
import com.google.gson.Gson
import java.io.InputStreamReader
import kotlin.system.measureTimeMillis

data class Result(val score: Int, val execTime: Long)

fun main() {
    val inputStream = Thread.currentThread().contextClassLoader.getResourceAsStream("trained-brain.json")
        ?: throw IllegalArgumentException("Resource not found")
    val reader = InputStreamReader(inputStream)
    val brainData: SnakeBrainData = Gson().fromJson(reader, SnakeBrainData::class.java)
    reader.close()

    val snakeGame = SnakeGame(Options(20, 20, null))
    val inputLayer = InputLayer(snakeGame)
    val snakeBrain = SnakeBrain(
        com.example.snake.ai.Options(
            brainData.inputLength,
            brainData.layerShapes,
            brainData.hiddenLayerActivationFunction,
            ProvidedWeightsAndBiases(
                brainData.weights.map { layer -> layer.map { it.clone() } },
                brainData.biases.map { it.clone() }
            )
        )
    )

    val games = 1000
    val result = mutableListOf<Result>()

    for (i in 0 until games) {
        snakeGame.reset()

        val execTime = measureTimeMillis {
            while (!snakeGame.gameOver) {
                val input = inputLayer.compute()
                val direction = snakeBrain.compute(input)
                snakeGame.snakeMoveByDirection(direction)
            }
        }

        result.add(Result(snakeGame.snake.length, execTime))
    }

    val totalTime = result.sumOf { it.execTime }.toDouble() / 1000
    println("Total time: ${"%.3f".format(totalTime)}s")

    val bestScore = result.maxByOrNull { it.score }?.score ?: 0
    println("Best score: $bestScore")

    val avgScore = result.sumOf { it.score } / games
    println("Average score: $avgScore")
}